File: .old/src_old/Lexer.lua
local Token = require("Token")

--- @class Lexer
--- @field _compiler Compiler
--- @field _token Token
local Lexer = {}
Lexer.__index = Lexer

Lexer.keywordSet = {
	["nil"] = true,
	["true"] = true,
	["false"] = true,
	["Nil"] = true,
	["Bool"] = true,
	["Real"] = true,
	["String"] = true,
	["def"] = true,
}

Lexer.punctuationSet = {
	["("] = true,
	[")"] = true,
	["{"] = true,
	["}"] = true,
	[","] = true,
	["="] = true,
	["!"] = true,
	["+"] = true,
	["-"] = true,
	["*"] = true,
	["/"] = true,
	["%"] = true,
	["<"] = true,
	[">"] = true,
	["=="] = true,
	["!="] = true,
	["<="] = true,
	[">="] = true,
	["||"] = true,
	["&&"] = true,
}

--- @param compiler Compiler
--- @return Lexer lexer
function Lexer.new(compiler)
	return setmetatable({
		_compiler = compiler,
		_token = Token.new("eof"),
	}, Lexer)
end

function Lexer:getCompiler()
	return self._compiler
end

function Lexer:getToken()
	if self._kind then
		return self._token
	else
		return self:nextToken()
	end
end

function Lexer:nextToken()
	local reader = self._compiler:getReader()
	local chr = reader:getChar()
	self._kind = nil -- eof
	self._value = nil
	if chr == "#" then -- comment
		-- skip line
		reader:consumeChars("[^\n]")
		reader:nextChar()
		return self:nextToken()
	elseif chr:find("%s") then -- whitespace
		reader:consumeChars("%s")
		return self:nextToken()
	elseif chr:find("[%a_$]") then -- names and keywords
		self._compiler:pushContext()
		self:_scanNamed()
		self._compiler:popContext()
	elseif chr:find("[%d%.]") then -- real literal
		self._compiler:pushContext()
		self:_scanRealLiteral()
		self._compiler:popContext()
	elseif chr == '"' or chr == "'" then -- string literal
		self._compiler:pushContext()
		self:_scanStringLiteral()
		self._compiler:popContext()
	elseif chr:find("%p") then -- punctuation
		self._compiler:pushContext()
		self:_scanPunctuation()
		self._compiler:popContext()
	end
	return self._token
end

--- Precondition: current character is a letter, underscore, or intrinsic specifier ("$").
function Lexer:_scanNamed()
	local reader = self._compiler:getReader()
	local intrinsic = reader:getChar() == "$"

	if intrinsic then
		-- skip intrinsic specifier
		reader:nextChar()
	end

	local lexeme = reader:consumeAndQuery("[%w_]")
	local chr = reader:getChar()

	if intrinsic then -- intrinsic
		self._kind = "<intrinsic>"
		self._value = lexeme
	elseif Lexer.keywordSet[lexeme] then -- keyword
		self._kind = lexeme
	else -- identifier
		self._kind = "<identifier>"
		self._value = lexeme
	end

	if chr == "$" then
		self._compiler:addDiagnostic(
			"warning",
			("consider adding a space after %q"):format(intrinsic and "$" .. lexeme or lexeme)
		)
	end
end

--- Precondition: current character is a digit or decimal point.
function Lexer:_scanRealLiteral()
	local reader = self._compiler:getReader()
	local lexeme = reader:consumeAndQuery("[%d%.]")
	local value = tonumber(lexeme)
	self._kind = "<real>"
	self._value = value

	-- ensure the real literal is valid
	if not value then
		self._compiler:addDiagnostic("error", ("malformed real literal %q"):format(lexeme))
		self._value = 0
	end

	-- warn for "weird" token sequence (e.g., 2x, 2.x, 2$)
	local chr = reader:getChar()
	if chr ~= "" and chr:find("[%w$]") then
		self._compiler:addDiagnostic("warning", ("consider adding a space after %q"):format(lexeme))
	end
end

--- Precondition: current character is a quote.
function Lexer:_scanStringLiteral()
	local reader = self._compiler:getReader()
	local quote = reader:getChar()

	-- consume quote
	reader:nextChar()

	-- consume non-quote
	local lexeme = reader:consumeAndQuery("[^" .. quote .. "]")
	self._kind = "<string>"
	self._value = lexeme

	-- expect quote
	if reader:getChar() == quote then
		reader:nextChar()
	else
		self._compiler:addDiagnostic("error", "unterminated string")
	end
end

--- Precondition: current character is punctuation.
function Lexer:_scanPunctuation()
	local reader = self._compiler:getReader()
	local chr = reader:getChar()

	-- greedily match operators
	local op2 = chr .. reader:nextChar()
	if #op2 == 2 and Lexer.punctuationSet[op2] then
		-- match operators of length 2 first
		reader:nextChar()
		self._kind = op2
	elseif Lexer.punctuationSet[chr] then
		-- match operators of length 1 next
		self._kind = chr
	else
		-- unexpected character, recover by scanning the next token
		self._compiler:addDiagnostic("error", ("unexpected character %q"):format(chr))
		return self:nextToken()
	end
end

return Lexer

File: .newer/lexer.py
from string import ascii_letters, digits, punctuation, whitespace
from typing import Callable, Tuple

Token = Tuple[str, str]


class Lexer:
    def __init__(self, source: str):
        self.source = source
        self.position = 0
        self.line = 1
        self.column = 1
        self.token = self.__scan_impl()

    def scan(self):
        self.token = self.__scan_impl()
        return self.token

    __alnum_set = set(ascii_letters + digits)
    __alpha_set = set(ascii_letters)
    __digit_set = set(digits)
    __punct_set = set(punctuation)
    __space_set = set(whitespace)

    def __scan_impl(self) -> Token:
        chr = self.__get_char()
        if chr == '':  # eof
            return ('eof', '')
        elif chr in self.__space_set:  # whitespace
            self.__next_char_while(lambda chr: chr in self.__space_set)
            return self.__scan_impl()
        elif chr == '#':  # comment
            self.__next_char_while(lambda c: c != '\n')
            self.__next_char()
            return self.__scan_impl()
        elif chr in self.__alpha_set or chr == '$' or chr == '_':  # keyword/identifier
            token = self.__scan_text()
            return token
        elif chr in self.__digit_set or chr == '.':  # real literal
            return self.__scan_real_literal()
        elif chr == '"' or chr == "'":  # string literal
            return self.__scan_string_literal()
        elif chr in self.__punct_set:  # punctuation
            return self.__scan_punctuation()
        else:
            raise ValueError('unexpected character', chr)

    def __get_char(self):
        return self.source[self.position] if self.position < len(self.source) else ''

    def __next_char(self):
        if self.position < len(self.source):
            chr = self.__get_char()
            if chr == '\n':
                self.line += 1
                self.column = 1
            else:
                self.column += 1
            self.position += 1
        return self.__get_char()

    def __next_char_while(self, predicate: Callable[[str], bool]):
        while self.position < len(self.source) and predicate(self.__get_char()):
            self.__next_char()
        return chr

    __keywords = {
        'def': True,
        'nil': True,
        'true': True,
        'false': True,
    }

    def __scan_text(self):
        start = self.position
        self.__next_char_while(
            lambda chr: chr in self.__alnum_set or chr == '$' or chr == '_')
        name = self.source[start:self.position]
        if name in self.__keywords:
            return (name, name)
        else:
            return ('<identifier>', name)

    def __scan_real_literal(self):
        start = self.position
        self.__next_char_while(lambda chr: chr in self.__digit_set)
        if self.__get_char() == '.':
            self.__next_char()
            self.__next_char_while(lambda chr: chr in self.__digit_set)
        lexeme = self.source[start:self.position]
        if lexeme == '.':
            raise ValueError('invalid real literal', lexeme)
        return ('<real>', self.source[start:self.position])

    def __scan_string_literal(self):
        start = self.position
        quote = self.__get_char()
        self.__next_char()
        self.__next_char_while(lambda chr: chr != quote)
        if self.__get_char() != quote:
            raise ValueError('expected character', quote)
        self.__next_char()
        return ('<string>', self.source[start + 1:self.position - 1])

    __punctuation = {
        '(': True, ')': True,
        '{': True, '}': True,
        ',': True,
        '=': True,
        '!': True,
        '+': True, '-': True, '*': True, '/': True, '%': True,
        '<': True, '>': True, '==': True, '!=': True, '<=': True, '>=': True,
        '||': True, '&&': True,
    }

    def __scan_punctuation(self):
        op1 = self.__get_char()
        op2 = op1 + self.__next_char()
        if op1 != op2 and op2 in self.__punctuation:
            self.__next_char()
            return (op2, op2)
        elif op1 in self.__punctuation:
            return (op1, op1)
        else:
            raise ValueError('unexpected character', op1)
